tests/branch_management/tests/test_branches.py

```diff
@@ -1,93 +1,127 @@
     patch_dir = Path("build-scripts/patches")
     branch = "HEAD" if branch is None else branch
-    cmd = f"git ls-tree --full-tree -r --name-only {branch} {patch_dir}"
-    output = check_output(cmd.split()).decode("utf-8")
+    cmd = f"git ls-tree --full-tree -r --name-only origin/{branch} {patch_dir}"
+    output = _sh(cmd.split())
     patches = set(
         Path(f).relative_to(patch_dir).parents[0] for f in output.splitlines()
     )
     return [p.name for p in patches]
 
 
-def _confirm_recipe(track, flavour):
+@functools.lru_cache
+def _confirm_branch_exists(repo, branch):
+    log.info(f"Checking {branch} branch exists in {repo}")
+    cmd = f"git ls-remote --heads {repo} {branch}"
+    output = _sh(cmd.split())
+    return branch in output
+
+
+def _confirm_all_branches_exist(leader):
+    assert _confirm_branch_exists(
+        K8S_GH_REPO, leader
+    ), f"GH Branch {leader} does not exist"
+    branches = [leader]
+    branches += [f"autoupdate/{leader}-{fl}" for fl in _branch_flavours(leader)]
+    if missing := [b for b in branches if not _confirm_branch_exists(K8S_GH_REPO, b)]:
+        assert missing, f"GH Branches do not exist {missing}"
+    if missing := [b for b in branches if not _confirm_branch_exists(K8S_LP_REPO, b)]:
+        assert missing, f"LP Branches do not exist {missing}"
+
+
+@functools.lru_cache
+def _confirm_recipe_exist(track, flavour):
     recipe = f"https://launchpad.net/~containers/k8s/+snap/k8s-snap-{track}-{flavour}"
     r = requests.get(recipe)
     return r.status_code == 200
 
 
-def test_branches(upstream_release):
-    """Ensures git branches exist for prior releases.
+def _confirm_all_recipes_exist(track, branch):
+    log.info(f"Checking {track} recipe exists")
+    assert _confirm_branch_exists(
+        K8S_GH_REPO, branch
+    ), f"GH Branch {branch} does not exist"
+    flavours = ["classic"] + _branch_flavours(branch)
+    recipes = {flavour: _confirm_recipe_exist(track, flavour) for flavour in flavours}
+    if missing := [fl for fl, exists in recipes.items() if not exists]:
+        assert missing, f"LP Recipes do not exist for {track} {missing}"
+
 
-    We need to make sure the LP builders pointing to the main github branch are only pushing
-    to the latest and current k8s edge snap tracks. An indication that this is not enforced is
-    that we do not have a branch for the k8s release for the previous stable release. Let me
-    clarify with an example.
+def test_prior_branches(prior_stable_release):
+    """Ensures git branches exist for prior stable releases.
 
-    Assuming upstream stable k8s release is v1.12.x, there has to be a 1.11 github branch used
-    by the respective LP builders for building the v1.11.y.
+    This is to ensure that the prior release branches exist in the k8s-snap repository
+    before we can proceed to build the next release. For example, if the current stable
+    k8s release is v1.31.0, there must be a release-1.30 branch before updating main.
```

Comment:
What happens if we explicitly don't want a release for a specific flavor? Maybe moonray should not have a 1.32 release (for whatever reason). Will this break all releases/checks for this flavor going forward?